include/aws/mqtt/private/request-response/protocol_adapter.h

```diff
@@ -0,0 +1,196 @@
+ * client.
+ */
+struct aws_protocol_adapter_subscription_event {
+    struct aws_byte_cursor topic_filter;
+    enum aws_protocol_adapter_subscription_event_type event_type;
+};
+
+/*
+ * An event emitted by the protocol adapter whenever a publish is received by the protocol client.  This will
+ * potentially include messages that are completely unrelated to MQTT request-response.  The topic is the first
+ * thing that should be checked for relevance.
+ */
+struct aws_protocol_adapter_incoming_publish_event {
+    struct aws_byte_cursor topic;
+    struct aws_byte_cursor payload;
+};
+
+/*
+ * Describes the type of connection event emitted by the protocol adapter
+ */
+enum aws_protocol_adapter_connection_event_type {
+    AWS_PACET_OFFLINE,
+    AWS_PACET_ONLINE,
+};
+
+/*
+ * An event emitted by the protocol adapter whenever the protocol client encounters a change in connectivity state.
+ */
+struct aws_protocol_adapter_connection_event {
+    enum aws_protocol_adapter_connection_event_type event_type;
+    bool rejoined_session;
+};
+
+typedef void(
+    aws_protocol_adapter_subscription_event_fn)(struct aws_protocol_adapter_subscription_event *event, void *user_data);
+typedef void(aws_protocol_adapter_incoming_publish_fn)(
+    struct aws_protocol_adapter_incoming_publish_event *publish,
+    void *user_data);
+typedef void(aws_protocol_adapter_terminate_callback_fn)(void *user_data);
+typedef void(
+    aws_protocol_adapter_connection_event_fn)(struct aws_protocol_adapter_connection_event *event, void *user_data);
+
+/*
+ * Set of callbacks invoked by the protocol adapter.  These must all be set.
+ */
+struct aws_mqtt_protocol_adapter_options {
+    aws_protocol_adapter_subscription_event_fn *subscription_event_callback;
+    aws_protocol_adapter_incoming_publish_fn *incoming_publish_callback;
+    aws_protocol_adapter_terminate_callback_fn *terminate_callback;
+    aws_protocol_adapter_connection_event_fn *connection_event_callback;
+
+    /*
+     * User data to pass into all singleton protocol adapter callbacks.  Likely either the request-response client
+     * or the subscription manager component of the request-response client.
+     */
+    void *user_data;
+};
+
+struct aws_mqtt_protocol_adapter_vtable {
+
+    void (*aws_mqtt_protocol_adapter_delete_fn)(void *);
+
+    int (*aws_mqtt_protocol_adapter_subscribe_fn)(void *, struct aws_protocol_adapter_subscribe_options *);
+
+    int (*aws_mqtt_protocol_adapter_unsubscribe_fn)(void *, struct aws_protocol_adapter_unsubscribe_options *);
+
+    int (*aws_mqtt_protocol_adapter_publish_fn)(void *, struct aws_protocol_adapter_publish_options *);
+};
+
+struct aws_mqtt_protocol_adapter {
+    const struct aws_mqtt_protocol_adapter_vtable *vtable;
+    void *impl;
+};
+
+AWS_EXTERN_C_BEGIN
+
+/*
+ * Creates a new request-response protocol adapter from an MQTT311 client
+ */
+AWS_MQTT_API struct aws_mqtt_protocol_adapter *aws_mqtt_protocol_adapter_new_from_311(
```

Comment:
Trivial: We usually stick to the format of `_new()` and pair it with `_destroy()`. Just wondering if this is different for any particular reason. e.g. why aren't we using `aws_mqtt_protocol_adapter_from_311_new()` and `aws_mqtt_protocol_adapter_destroy()`?

Reply:
Pattern is <subject><verb><qualifiers>

For ex: https://github.com/awslabs/aws-c-auth/blob/main/include/aws/auth/credentials.h#L650-L1116

Reply:
Ah, gotcha. I don't think I've seen the qualifiers around till now.